controllers/provisioners/eksfargate/eksfargate.go

```diff
@@ -0,0 +1,290 @@
+	}
+	return stringReferences
+}
+func CreateFargateTags(tagArray []map[string]string) map[string]*string {
+	tags := make(map[string]*string)
+	for _, t := range tagArray {
+		for k, v := range t {
+			va := new(string)
+			*va = v
+			tags[k] = va
+		}
+	}
+	return tags
+}
+
+// Convienence function to convert from json to API.
+func CreateFargateSelectors(selectors []v1alpha1.EKSFargateSelectors) []*eks.FargateProfileSelector {
+	var eksSelectors []*eks.FargateProfileSelector
+	for _, selector := range selectors {
+		m := make(map[string]*string)
+		for k, v := range selector.Labels {
+			vv := new(string)
+			*vv = v
+			m[k] = vv
+		}
+		eksSelectors = append(eksSelectors, &eks.FargateProfileSelector{Namespace: &selector.Namespace, Labels: m})
+	}
+	return eksSelectors
+}
+func IsRetryable(instanceGroup *v1alpha1.InstanceGroup) bool {
+	for _, state := range NonRetryableStates {
+		if state == instanceGroup.GetState() {
+			return false
+		}
+	}
+	return true
+}
+func (ctx *InstanceGroupContext) Create() error {
+	var arn string
+	instanceGroup := ctx.GetInstanceGroup()
+	spec := instanceGroup.GetEKSFargateSpec()
+	if instanceGroup.GetEKSFargateSpec().GetPodExecutionRoleArn() == "" {
+		created, err := ctx.AwsWorker.CreateDefaultRole()
+		if created || err != nil {
+			ctx.Log.Info("Creating default role")
+			return err
+		}
+		role, err := ctx.AwsWorker.GetDefaultRole()
+		if err != nil {
+			ctx.Log.Info("Failed to get default role", "error", err)
+			return err
+		}
+		arn = *role.Arn
+
+		err = ctx.AwsWorker.AttachDefaultPolicyToDefaultRole()
+		if err != nil {
+			ctx.Log.Info("Failed to get attach policy to role", "error", err)
+			return err
+		}
+		ctx.Log.Info("Attached default policy to role")
+
+	} else {
+		arn = spec.GetPodExecutionRoleArn()
+	}
+
+	ctx.Log.Info("Creating a profile.", "arn", arn)
+	tryAgain, err := ctx.AwsWorker.CreateProfile(arn)
+	if tryAgain {
+		ctx.Log.Info("Resource inuse on Create.")
+		return nil
+	}
+	if err != nil {
+		ctx.Log.Info("Creation of the fargate profile failed", "cluster",
+			spec.GetClusterName(),
+			"profile",
+			spec.GetProfileName(),
+			"error", err)
+	} else {
+		instanceGroup.SetState(v1alpha1.ReconcileModifying)
+	}
+
+	return err
+}
+func (ctx *InstanceGroupContext) CloudDiscovery() error {
+	profile, err := ctx.AwsWorker.DescribeProfile()
+	if err != nil {
+		profile = &eks.FargateProfile{
+			Status: nil,
+		}
+	}
+	if profile.Status == nil {
+		ctx.DiscoveredState.ProfileStatus = aws.FargateProfileStatusMissing
+	} else {
+		ctx.DiscoveredState.ProfileStatus = *profile.Status
+	}
+	return nil
+}
+func (ctx *InstanceGroupContext) Delete() error {
+	instanceGroup := ctx.GetInstanceGroup()
+	spec := instanceGroup.GetEKSFargateSpec()
+
+	worker := ctx.AwsWorker
+	if spec.GetPodExecutionRoleArn() == "" {
+		found, err := worker.DetachDefaultPolicyFromDefaultRole()
+		if found || err != nil {
+			ctx.Log.Info("Detaching the default policy", "error", err)
```

Comment:
Should you log an Error here instead of Info? can you also make sure the message is more descriptive e.g. "failed to detach policy". 
Also, should we return an error if `found` is true? given that this is the delete, won't this always return true at the first pass?

Reply:
1) found == true & no error -> implies it was successfully detached so return for a requeue.
2) found == true & error -> the detach failed for some systemic reason and we need to return the error upwards.
3) found == false & no error -> implies the detach has previously been done so proceed forward in the Delete() logic to the next resource
4) found == false & error -> this case does not exist

so the logic 
`
if found || err != nil {
                          ctx.Log.Info("Detaching the default policy", "error", err)
                          return err
                  }
`
returns immediately when found==true and the error/no error bubbles up
found == false and no error proceeds forward to get rid of the next resource.

Reply:
I didn't want the "iam.ErrCodeNoSuchEntityException" to bubble up when calling the detach twice is the normal case.

Reply:
As mentioned above, we should make this logic specific to the provisioner, so you can send the error back to the caller (`Delete()`), and handle the specific case there.

This will enable you to simply return an error from `aws.go` methods and decided what to do when there is a specific error inside your provisioner

BTW, I know other methods in `aws.go` do the same thing, but we will be changing this in the future, most aws err handling should happen in the provisioner as that is where the specific logic is.

Reply:
OK and you are OK if I check the exception (iam.ErrCodeNoSuchEntityException) in the Delete()?  It will make provisioners/eksfargate/eksfargate.go more dependent upon AWS.

Reply:
Sure, if that's not a problem with imports, we can definitely do that,
It's also OK to make a more concrete method in `aws.go` and then handle that error there if it's more convenient. example `CreateIAMRole` would return error to caller, and `CreateIAMRoleIfNotExist` would ignore that error -> if we see multiple provisioners needing the same logic we should definitely do that.

Reply:
Hmmm.  Good idea, maybe later.